src/Security/Authentication/Negotiate/src/NegotiateHandler.cs

```diff
@@ -0,0 +1,334 @@
+                {
+                    persistence ??= EstablishConnectionPersistence(connectionItems);
+                    Debug.Assert(persistence.State == null, "Complete Kerberos results should only be produced from a new context.");
+                    persistence.State = _negotiateState;
+                }
+
+                // Note we run the Authenticated event in HandleAuthenticateAsync so it is per-request rather than per connection.
+            }
+            catch (Exception ex)
+            {
+                var errorContext = new AuthenticationFailedContext(Context, Scheme, Options) { Exception = ex };
+                await Events.AuthenticationFailed(errorContext);
+
+                if (errorContext.Result != null)
+                {
+                    if (errorContext.Result.Handled)
+                    {
+                        return true;
+                    }
+                    else if (errorContext.Result.Skipped)
+                    {
+                        return false;
+                    }
+                    else if (errorContext.Result.Failure != null)
+                    {
+                        throw new Exception("An error was returned from the AuthenticationFailed event.", errorContext.Result.Failure);
+                    }
+                }
+
+                throw;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Checks if the current request is authenticated and returns the user.
+        /// </summary>
+        /// <returns></returns>
+        protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
+        {
+            if (!_requestProcessed)
+            {
+                throw new InvalidOperationException("AuthenticateAsync must not be called before the UseAuthentication middleware.");
+            }
+
+            if (IsHttp2)
+            {
+                // Not supported. We don't throw because Negotiate may be set as the default auth
+                // handler on a server that's running HTTP/1 and HTTP/2. We'll challenge HTTP/2 requests
+                // that require auth and they'll downgrade to HTTP/1.1.
+                return AuthenticateResult.NoResult();
+            }
+
+            if (_negotiateState == null)
+            {
+                return AuthenticateResult.NoResult();
+            }
+
+            if (!_negotiateState.IsCompleted)
+            {
+                // This case should have been rejected by HandleRequestAsync
+                throw new InvalidOperationException("Attempting to use an incomplete authentication context.");
+            }
+
+            // Make a new copy of the user for each request, they are mutable objects and
+            // things like ClaimsTransformation run per request.
+            var identity = _negotiateState.GetIdentity();
+            ClaimsPrincipal user;
```

Comment:
Only use var when it's obvious what type of variable is being returned, e.g. when using new. I have no idea what type identity is by reading this code.